### ISSUE TITLE
[Bug] Triton mask-map moe_permute ignores num_out_tokens

### ISSUE THREAD

**Issue by truong-v** [NVIDIA/TransformerEngine#3346]
## Summary

The Triton path behind `moe_permute(..., map_type="mask")` sizes its output at exactly `num_out_tokens` rows but never enforces that limit: `_permute_kernel` receives `num_out_tokens` explicitly marked `# pylint: disable=unused-argument`, the row-id map assigns destination rows as an unbounded cumulative sum over the routing map, and nothing clamps them. If the routing map routes more tokens than `num_out_tokens`, the excess rows are written past the end of the allocation.

The CUDA path behind the same public API (`map_type="index"`) implements the capacity limit — `moe_permute_row_map` has an explicit `idx >= num_out_tokens` branch that marks over-capacity entries `-1`, and every consumer skips `-1` (`common/permutation/permutation.cu` lines 28–34) — so the two backends of one entry point disagree, and a caller expecting the documented capacity-drop behaviour from the mask path gets out-of-bounds writes. #2876 added a positivity guard for `num_out_tokens`, but a positive value smaller than `routing_map.sum()` still reaches the kernel unchecked, and the tests always construct the map so that `routing_map.sum() == num_out_tokens`, which is why this has not been seen.

The kernel was added in #1373.

## Steps to reproduce

```python
import torch
import transformer_engine.pytorch  # noqa
from transformer_engine.pytorch.permutation import moe_permute

routing_map = torch.zeros(8, 4, dtype=torch.int32, device="cuda")
routing_map[:, 0] = 1
routing_map[:4, 1] = 1          # 12 routed entries
inp = torch.randn(8, 16, device="cuda", dtype=torch.bfloat16)
out, row_id_map = moe_permute(inp, routing_map, num_out_tokens=9, map_type="mask")
# out has 9 rows; the kernel writes 12
```

Under `compute-sanitizer --tool memcheck` (with `PYTORCH_NO_CUDA_MEMORY_CACHING=1` so the caching allocator's slack does not absorb it), this reports 48 invalid global writes at the permute store (`transformer_engine/common/triton/permutation.py:303`) — the three over-capacity rows times the 16-element hidden size — and the run then ends in an unspecified launch failure. After the fix: 0 errors, the in-capacity rows are unchanged, and the row map's per-token `n_routed` sums to `num_out_tokens`.

## Environment

- TransformerEngine commit 07e281f2 (main); transformer-engine-cu12 2.17.0 wheel + source checkout for the Python/Triton code
- NVIDIA L40S (sm_89), driver CUDA 12.4
- torch 2.13.0+cu126, triton 3.7.1, Python 3.12

**Comment by modelpath-dev** [comment]
I will take this issue. Please assign it to me.

The problem is that the Triton path in `moe_permute` does not enforce the `num_out_tokens` limit, leading to out-of-bounds writes. I will first inspect the `_permute_kernel` function and the row-id map logic in `transformer_engine/common/triton/permutation.py`. The fix will involve adding a check to ensure that the cumulative sum of the routing map does not exceed `num_out_tokens`, similar to the CUDA path. I will verify the fix by running the provided reproduction steps and checking for the absence of invalid writes.

**Comment by modelpath-dev** [comment]
I have started looking into the `_permute_kernel` function and the row-id map logic. I will add a check to ensure that the cumulative sum of the routing map does not exceed `num_out_tokens`. Once I implement the fix, I will run the reproduction steps to verify that the issue is resolved and there are no invalid writes. Let me know if there are any additional considerations.
